### PR TITLE
patch(TS): canvas patch init 2

### PR DESCRIPTION
**Pull request by ShaMan123**
## Description

Refactor canvas init logic, more work can be done in this scope

## Changes

- canvas init
- dumped event binding tests since we bind in constructor from now on
- remove `requestRenderAllBound`, `renderAndResetBound`
- ~~fix `_initRetinaScaling` regression from #8510, BREAKING since now should be called only if `isRetinaScaling` (doesn't check in the method as previously did)~~

### CODE REVIEW

File: src/canvas.class.ts [hunk 1]
```diff
@@ -501,20 +501,16 @@ export class Canvas<
    */
   constructor(el: string | HTMLCanvasElement, options = {}) {
     super(el, options);
+    // @ts-ignore
+    this._initEventListeners();
+    this._createCacheCanvas();
```
Comment (ShaMan123) [inline src/canvas.class.ts:506]: should this belong to `initElements`? Or does it relay on retina scaling being initialized?

File: src/static_canvas.class.ts [hunk 2]
```diff
@@ -278,6 +275,22 @@ export class StaticCanvas<
     kill: (reason?: any) => void;
   };
 
+  constructor(el: string | HTMLCanvasElement, options = {}) {
+    super();
+    this.renderAndReset = this.renderAndReset.bind(this);
+    this.requestRenderAll = this.requestRenderAll.bind(this);
```
Comment (ShaMan123) [inline src/static_canvas.class.ts:281]: renamed these

**Review by ShaMan123 — COMMENTED** [review]
READY

**Review by ShaMan123 — COMMENTED** [review]
updated from master

**Comment by ShaMan123** [comment]
now things are broken

**Comment by ShaMan123** [comment]
@asturur this is high priority 
I have no idea the full extent of the regression but to fix it I had to init size on upper canvas, something I didn't locate be #8510 
I then tried to use `setDimensions` instead of `initSize` with no luck.
please review

File: src/canvas/canvas.class.ts [hunk 3]
```diff
@@ -497,34 +497,16 @@ export class SelectableCanvas<
...
     this._createCacheCanvas();
   }
 
-  /**
-   * @private
-   */
-  _initRetinaScaling() {
+  protected _initRetinaScaling() {
     super._initRetinaScaling();
     this.__initRetinaScaling(this.upperCanvasEl, this.contextTop);
```
Comment (ShaMan123) [inline src/canvas/canvas.class.ts:511]: the regression was here
I changed this method to run w/o a check if is retina scaling and call it only if so

File: src/canvas/canvas.class.ts [hunk 4]
```diff
@@ -1252,30 +1234,15 @@ export class SelectableCanvas<
...
   }
 
-  /**
-   * @private
-   */
-  _createCacheCanvas() {
+  protected _createCacheCanvas() {
     this.cacheCanvasEl = this._createCanvasElement();
-    this.cacheCanvasEl.setAttribute('width', `${this.width}`);
-    this.cacheCanvasEl.setAttribute('height', `${this.height}`);
```
Comment (ShaMan123) [inline src/canvas/canvas.class.ts:1266]: size will be set by `setDimensions` on init

File: src/canvas/canvas_events.ts [hunk 5]
```diff
@@ -215,47 +224,6 @@ export class Canvas extends SelectableCanvas {
     );
   }
 
-  /**
-   * @private
-   */
-  private _bindEvents() {
```
Comment (ShaMan123) [inline src/canvas/canvas_events.ts:221]: no need to have this now that it is impossible to init canvas twice

File: src/canvas/static_canvas.class.ts [hunk 6]
```diff
@@ -271,10 +272,7 @@ export class StaticCanvas<
    * in order to return it to original state on dispose
    * @type string
    */
-  _originalCanvasStyle?: string;
-
-  renderAndResetBound: () => void;
-  requestRenderAllBound: () => void;
```
Comment (ShaMan123) [inline src/canvas/static_canvas.class.ts:277]: dropped

**Review by ShaMan123 — COMMENTED** [review]
ready

File: src/canvas/static_canvas.class.ts [hunk 7]
```diff
@@ -283,7 +281,24 @@ export class StaticCanvas<
 
   constructor(el: string | HTMLCanvasElement, options = {}) {
     super();
-    this._init(el, options);
+    this.renderAndReset = this.renderAndReset.bind(this);
+    this.requestRenderAll = this.requestRenderAll.bind(this);
```
Comment (ShaMan123) [inline src/canvas/static_canvas.class.ts:285]: these were renamed,
see description

**Review by ShaMan123 — COMMENTED** [review]
mergable

File: src/canvas/static_canvas.class.ts (hunk 7 above)
Comment (asturur) [reply src/canvas/static_canvas.class.ts:285]: Yes yes i have read everything i was in the middle of things and then i had to go to bed.
Probably going to let them both names.

**Comment by asturur** [comment]
Bind on constructor is a great normal thing we should have done as soon as we started moving canvas, i really didn't thin of it.
I don't want to have requestRenderAll and renderAndReset naturally bound for those reasons:
- are not meant to be callbacks or event handlers
- an unexperienced dev would see them working and not understand those need to be bound
- you could override them or swap them on the prototype and then assign back to the instance and loose the binding
So i delete the convenience bound functions ( with the bound suffix ) but i m not binding this.
If there is a complain that the old one were useful we can add them back

The reorganization is great, i don't like the idea of adding a parameter to setDimensions to have it skip rendering just because we need it for initialization, since we managed to remove eventsBound: boolean, let's not add a similar one.
I separated setDimenions in setDimensions + protected setDimensionsImplementation, and teh constructor can call the latter.

**Review by ShaMan123 — COMMENTED** [review]
I absolutely agree

**Comment by ShaMan123** [comment]
@asturur why e1ffe0cac797d20c8c76ac09bc8d77d104d55fb2?

**Comment by asturur** [comment]
it is how it was before and i wouldn't mind if it doesn't save half kb on the minified build.

**Review by asturur — APPROVED** [review]

**Pull request merged**